crypto/fipsmodule/rand/snapsafe_detect.c

```diff
@@ -0,0 +1,126 @@
+// -----------------------------------------------------------------------------
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+// -----------------------------------------------------------------------------
+
+#include <openssl/crypto.h>
+
+#include "snapsafe_detect.h"
+
+#if defined(OPENSSL_LINUX)
+
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#include "../delocate.h"
+
+DEFINE_STATIC_ONCE(g_snapsafe_detect_once)
+DEFINE_STATIC_MUTEX(g_snapsafe_detect_lock)
+DEFINE_BSS_GET(volatile uint32_t *, g_sysgenid_addr)
+DEFINE_BSS_GET(int, g_ignore_snapsafe)
+DEFINE_BSS_GET(int, g_overwrite_sysgenid)
+
+static void init_snapsafe_detect(void) {
+
+  // For testing purposes, we sometimes ignore snapsafe detection or overwrite
+  // the SysGenID callback.
+  if (*g_ignore_snapsafe_bss_get() == 1 || *g_overwrite_sysgenid_bss_get() == 1) {
+    return;
+  }
+
+  long page_size = sysconf(_SC_PAGESIZE);
+  if (page_size <= 0) {
+    return;
+  }
+
+  int fd_sysgenid = open(AWSLC_SYSGENID_FILE_PATH, O_RDONLY);
+  if (fd_sysgenid == -1) {
+    return;
+  }
+
+  void *addr = mmap(NULL, (size_t) page_size, PROT_READ, MAP_SHARED,
+                    fd_sysgenid, 0);
+
+  // Can close |fd_sysgenid| now per
+  // https://man7.org/linux/man-pages/man2/mmap.2.html: "After the mmap() call
+  // has returned, the file descriptor, |fd_sysgenid|, can be closed immediately
+  // without invalidating the mapping.". We have initialised Snapsafe detection
+  // without errors and |init_snapsafe_detect| is only called once. Therefore,
+  // try to close |fd_sysgenid|, but don't error if it fails.
+  close(fd_sysgenid);
+
+  if (addr == MAP_FAILED) {
+    return;
+  }
+
+  *g_sysgenid_addr_bss_get() = addr;
+}
+
+int CRYPTO_get_snapsafe_generation(uint32_t *snapsafe_generation_number) {
+
+  // Best-effort attempt to initialise Snapsafe detection.
+  CRYPTO_once(g_snapsafe_detect_once_bss_get(), init_snapsafe_detect);
+
+  volatile uint32_t *const sysgenid_addr = *g_sysgenid_addr_bss_get();
+  if (sysgenid_addr == NULL) {
+    // SysGenId is not supported on system.
+    return 0;
+  }
+
+  struct CRYPTO_STATIC_MUTEX *const lock = g_snapsafe_detect_lock_bss_get();
+
+  // Aquire both read and write locks for a short while.
+  CRYPTO_STATIC_MUTEX_lock_write(lock);
+  *snapsafe_generation_number = *sysgenid_addr;
+  CRYPTO_STATIC_MUTEX_unlock_write(lock);
```

Comment:
why do we need locks here? reading `u32` from memory is a natively atomic operation for 32-bit and 64-bit systems ; cache coherency should not be an issue either since `sysgenid_addr` is a `volatile` pointer, right?

Reply:
I don't understand the reasons behind the bullet points. I was looking at fork_detect.c and I see the mutex is used there due to the multi-threading concern. Is there no concern for multi-threading here?

Reply:
The fork detection code maintains the fork generation number using shared memory. In particular, the memory pointed to by `g_fork_detect_addr`, defined through the macro:
```
DEFINE_BSS_GET(volatile char *, g_fork_detect_addr)
```
`CRYPTO_get_fork_generation` mutates this shared memory and must therefore implement locks around read and writes.

The snapsafe detection code does not maintain the generation number, it is managed by the SysGenId device (or whatever controls the device). `CRYPTO_get_snapsafe_generation` reads the value of the device which is atomic (because it is a 32-bit value). `snapsafe_generation_number` is a pointer from the calling code. In all calls in AWS-LC it points to local memory and is therefore not shared across the process. If `snapsafe_generation_number` was shared memory, the caller is better positioned to synchronise.

`CRYPTO_get_snapsafe_generation` does not mutate the SysGenID value either.

Reply:
In fork_detect.c there are multiple operations that need to _run together atomically_, mutex protects a sequence of operations there.

In this line the mutex doesn't add value, only overhead. This single line is thread-safe by itself:
- reading `*sysgenid_addr` can either give old or new value, never a mix, adding the mutex doesn't change anything here, right?
- `snapsafe_generation_number` points to local variable on the stack, each thread has its own copy, no contention on writing.

For completeness, I would add a documentation to the function that the function expects exclusive write access to memory pointed by `uint32_t *snapsafe_generation_number` parameter. If the caller wishes to use some shared memory/variable as the destination, they need to synchronize access outside of this function.

Reply:
Thank you both for the detailed explanation.